### PR TITLE
[FEAT] Visiting players land in Land Expansion

### PR DESCRIPTION
**Pull request by spencerdezartsmith**
# Description

Co-Authored: @Rofrtd 

This PR adds the ability to visit a land in land expansion. A player will be able to load a land to visit either by direct url `/visit/:id` or via the in game Island Travel system. Once visiting you can continue to visit other lands or you can return back to your game (if you used the in game travel) or back to the home screen (visit via url).

When visiting you will be able to see the current state of the land you're visiting including bumpkin details and inventory. There will be no game events available.

Clicking on the Bumpkin HUD will show the same content so you can check out folks' bumpkins but the `Visit bumpkin` link will take you to OpenSea instead.

Only lands that have migrated to Sunflower Isles can be visited. We are phasing out the current games visiting system. You will still be able to visit a contributor in the current game but thats it. We will move that over once we fully migrate.

Screen Shot 2022-11-03 at 11 09 34 am
Screen Shot 2022-11-03 at 11 09 49 am
Screen Shot 2022-11-03 at 11 10 20 am
Screen Shot 2022-11-03 at 11 12 21 am
Screen Shot 2022-11-03 at 11 12 41 am

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

**NOTE** I have migrated Land #30 and Land #4 on testnet so these farms can be visited.

- If you run locally against your own instance you will need to modify a land to have the `migrated: true` flag in order to visit that land.
- Please test visiting via `/visit/:id` and also via the Travel System in LE.
- Please also run the current game to make sure all is groovy 🕺🏼 

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [ ] I have read the contributing guidelines and agree to the T&Cs
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/features/game/expansion/Game.tsx [hunk 1]
```diff
@@ -113,6 +122,78 @@ export const Game: React.FC = () => {
     );
   }
 
+  if (gameState.matches("landToVisitNotFound")) {
+    return (
```
Comment (adamhannigan) [inline src/features/game/expansion/Game.tsx:126]: Consider moving this component elsewhere to keep Game.tsx focussed on the high level components.

File: src/features/game/expansion/components/IslandList.tsx [hunk 2]
```diff
@@ -92,25 +123,69 @@ export const IslandList = ({ bumpkin }: { bumpkin: Bumpkin | undefined }) => {
...
   ];
 
+  const handleBackToHomeScreen = () => send("RETURN");
+
   const islandList = islands
     .filter((item) => item.path !== location.pathname)
     .sort((a, b) => (a.levelRequired > b.levelRequired ? 1 : -1));
 
+  // Someone who is visiting without a loaded session
+  const unAuthenticatedVisit = authState.matches("visiting");
```
Comment (adamhannigan) [inline src/features/game/expansion/components/IslandList.tsx:140]: Can you be authenticated and also visiting?

File: src/features/game/expansion/components/IslandList.tsx [hunk 3]
```diff
@@ -69,9 +73,36 @@ const Island = ({ name, levelRequired, path, bumpkin, image }: Props) => {
...
+  showVisitList,
+}: {
+  bumpkin: Bumpkin | undefined;
+  showVisitList: boolean;
+}) => {
+  const { authService } = useContext(Auth.Context);
+  const [authState, send] = useActor(authService);
   const { id } = useParams();
   const location = useLocation();
+  const [view, setView] = useState<"list" | "form">("list");
```
Comment (adamhannigan) [inline src/features/game/expansion/components/IslandList.tsx:104]: What is a form for the IslandList? If this is visiting related can we make this a bit more specific.

E.g. `visitForm`

**Review by adamhannigan — COMMENTED** [review]
LGTM. Just a few suggestions around keeping the `Game.tsx` focussed.

I haven't manually tested this yet, waiting for the go ahead on the API ✅

File: src/features/game/expansion/components/IslandList.tsx (hunk 2 above)
Comment (spencerdezartsmith) [reply src/features/game/expansion/components/IslandList.tsx:140]: Yes you can but this state `authState.matches("visiting")` would not represent this. This is the `visiting` state on the `authMachine`. If you enter this state it means you've come via a url or you've clicked the `Explore a farm` button on the splash page. You have never loaded a session.

File: src/features/game/expansion/Game.tsx (hunk 1 above)
Comment (spencerdezartsmith) [reply src/features/game/expansion/Game.tsx]: I have moved this down a little to be a conditional render of the `GameContent` function. There is really no other place for this to go. If a player trying to visit and no land is found then this rendered component IS the Game in that instance. This is basically the `visiting` state when no land is found. It has its on hud etc.

**Comment by adamhannigan** [comment]
All worked well 👏 

I just found one small bug if I entered the URL manually into the URL bar it would not find the farm 🤔 If I type it in the form, it works though

Screen Shot 2022-11-03 at 4 50 55 pm

**Comment by spencerdezartsmith** [comment]
> All worked well 👏
> 
> I just found one small bug if I entered the URL manually into the URL bar it would not find the farm 🤔 If I type it in the form, it works though
> 
> Screen Shot 2022-11-03 at 4 50 55 pm

I think you just need to highlight the url and hit enter again. It just wasn't reloading the page.

It is working fine for me.

**Pull request merged**